{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def out_map(out_pop, chrom):\n",
    "\n",
    "    ###### Create the new allel map from phase1 to phase2 ######\n",
    "\n",
    "    pos_phase1 = allel.SortedIndex(calldata_phase1[chrom][\"variants/POS\"][:])\n",
    "    pos_phase2 = allel.SortedIndex(calldata_phase2[chrom][\"variants/POS\"][:])\n",
    "    loc1, loc2 = pos_phase2.locate_intersection(pos_phase1)\n",
    "    pos_p2_sel = pos_phase2.compress(loc1)\n",
    "    pos_p1_sel = pos_phase1.compress(loc2)\n",
    "    variants_phase1 = allel.VariantChunkedTable(calldata_phase1[chrom][\"variants\"], \n",
    "                                     names=['POS', 'REF', 'ALT', 'DP', 'MQ', 'QD', 'numalt'],\n",
    "                                     index='POS')\n",
    "    variants_phase1_filt = variants_phase1.compress(loc2, axis=0)\n",
    "    phase1_ref = variants_phase1_filt[\"REF\"][:]\n",
    "    phase1_alt = variants_phase1_filt[\"ALT\"][:]\n",
    "    variants_phase2 = allel.VariantChunkedTable(calldata_phase2[chrom][\"variants\"], \n",
    "                                     names=['POS', 'REF', 'ALT', 'DP', 'MQ', 'QD', 'numalt'],\n",
    "                                     index='POS')\n",
    "    variants_phase2_filt = variants_phase2.compress(loc1, axis=0)\n",
    "    phase2_ref = variants_phase2_filt[\"REF\"][:]\n",
    "    phase2_alt = variants_phase2_filt[\"ALT\"][:]\n",
    "    phase2refalt = np.column_stack([phase2_ref, phase2_alt])\n",
    "    mapping = allel.create_allele_mapping(phase1_ref, phase1_alt, phase2refalt)\n",
    "    \n",
    "    ###### Now Mapping on our selected Outgroup ######\n",
    "    \n",
    "    \n",
    "    calldata_outgroup= h5py.File('/gcs/phase1.AR3/extras/outgroup_allele_counts.h5', mode='r')\n",
    "    calldata_out_pop = calldata_outgroup[chrom][out_pop]\n",
    "    ac_out_pop = allel.AlleleCountsArray(calldata_out_pop)\n",
    "    ac_out_pop = ac_out_pop.compress(loc2)\n",
    "    pop_map_ac = ac_out_pop.map_alleles(mapping)\n",
    "\n",
    "    return pop_map_ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def abba_baba_plot(chrom, a, b, c, out_pop, windows_size):\n",
    "\n",
    "    ###### loading phase2 metadata ######\n",
    "    metadata = pd.read_csv(\"/home/jovyan/notebooks/samples.meta.txt\", sep=\"\\t\")\n",
    "    pop_select = metadata.population.isin({a, b, c}).values\n",
    "    pop_subset = metadata[pop_select]\n",
    "    \n",
    "    ###### loading phase2 genome and subset ######\n",
    "    pos_phase1 = allel.SortedIndex(calldata_phase1[chrom][\"variants/POS\"][:])\n",
    "    pos_phase2 = allel.SortedIndex(calldata_phase2[chrom][\"variants/POS\"][:])\n",
    "    loc1, loc2 = pos_phase2.locate_intersection(pos_phase1)\n",
    "    pos_p2_sel = pos_phase2.compress(loc1)\n",
    "    genotypes_phase2_call = calldata_phase2[chrom][\"calldata/GT\"]\n",
    "    genotypes_phase2 = allel.GenotypeChunkedArray(genotypes_phase2_call)\n",
    "    geno_p2_subset = genotypes_phase2.subset(sel0=loc1, sel1=pop_select)\n",
    "\n",
    "    \n",
    "    ###### perform allele count to my subpopulations ######\n",
    "    grp = pop_subset.groupby(\"population\")\n",
    "    grp_indices = grp.indices\n",
    "    ac_subpops = geno_p2_subset.count_alleles_subpops(grp_indices)\n",
    "    \n",
    "    ###### perform abba baba test ######\n",
    "    y = allel.average_patterson_d(ac_subpops[a], ac_subpops[b], ac_subpops[c], out_pop, windows_size)\n",
    "    print ('Estimated value of the statistic using all data:' ,y[0])\n",
    "    print ('Estimated standard error:' ,y[1])\n",
    "    print ('Z-score:' ,y[2])\n",
    "    \n",
    "    import csv\n",
    "    with open('file.csv', 'a') as csvFile:\n",
    "        tbl = [y[0],y[1],y[2]]\n",
    "        writer = csv.writer(csvFile)\n",
    "        writer.writerow(tbl)\n",
    "        csvFile.close()\n",
    "    \n",
    "    ###### compute windows with equal numbers of SNPs ######\n",
    "    windows = allel.moving_statistic(pos_p2_sel, statistic=lambda v: [v[0], v[-1]], size=windows_size)\n",
    "    x = np.asarray(windows).mean(axis=1)\n",
    "    \n",
    "    ###### Plot the test ######\n",
    "    fig, ax = pyplot.subplots(figsize=(12, 4))\n",
    "    sns.despine(ax=ax, offset=10)\n",
    "    ax.plot(x, y[3], lw=.5)\n",
    "    ax.set_ylabel(\"D Value\")\n",
    "    ax.set_xlabel('Chromosome %s position (bp)' %chrom)\n",
    "    ax.set_xlim(0, pos_p2_sel.max())\n",
    "    ax.set_title(('ABBA BABA Test between %s, %s, %s and outgroup population on chromosome %s') % (a,b,c,chrom))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pbs_plot(chrom, a, b, c, windows_size):\n",
    "\n",
    "    ###### loading phase2 metadata ######\n",
    "    metadata = pd.read_csv(\"/home/jovyan/notebooks/samples.meta.txt\", sep=\"\\t\")\n",
    "    pop_select = metadata.population.isin({a, b, c}).values\n",
    "    pop_subset = metadata[pop_select]\n",
    "    \n",
    "    ###### loading phase2 genome and subset ######\n",
    "    genotypes_phase2_call = calldata_phase2[chrom][\"calldata/GT\"]\n",
    "    genotypes_phase2 = allel.GenotypeChunkedArray(genotypes_phase2_call)\n",
    "    pos = allel.SortedIndex(calldata_phase2[chrom][\"variants/POS\"])\n",
    "    geno_p2_subset = genotypes_phase2.subset(sel1=pop_select)\n",
    "\n",
    "    \n",
    "    ###### perform allele count to my subpopulations ######\n",
    "    grp = pop_subset.groupby(\"population\")\n",
    "    grp_indices = grp.indices\n",
    "    ac_subpops = geno_p2_subset.count_alleles_subpops(grp_indices)\n",
    "    \n",
    "    ##### perform pbs #####\n",
    "    pbs = allel.pbs(ac_subpops[a], ac_subpops[b], ac_subpops[c], windows_size)\n",
    "    \n",
    "    ###### compute windows with equal numbers of SNPs ######\n",
    "    windows = allel.moving_statistic(pos, statistic=lambda v: [v[0], v[-1]], size=windows_size)\n",
    "    x = np.asarray(windows).mean(axis=1)\n",
    "    \n",
    "    ###### Plot the test ######\n",
    "    fig, ax = pyplot.subplots(figsize=(12, 4))\n",
    "    sns.despine(ax=ax, offset=10)\n",
    "    ax.plot(x, pbs, lw=.5)\n",
    "    ax.set_ylabel(\"PBS Value\")\n",
    "    ax.set_xlabel('Chromosome %s position (bp)' %chrom)\n",
    "    ax.set_xlim(0, pos.max())\n",
    "    ax.set_title(('PBS between %s, %s, %s populations on chromosome %s') % (a,b,c,chrom))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_dxy(pop1, pop2, chrom, window_size=20000, min_n_bases=1):\n",
    "    metadata = pd.read_csv(\"/home/jovyan/notebooks/samples.meta.txt\", sep=\",\")\n",
    "    pop_select = metadata.population.isin({pop1,pop2}).values\n",
    "    pop_subset = metadata[pop_select]\n",
    "    genotypes_phase2_call = calldata_hap_phase2[chrom][\"calldata/GT\"]\n",
    "    genotypes_phase2 = allel.GenotypeChunkedArray(genotypes_phase2_call)\n",
    "    geno_p2_subset = genotypes_phase2.subset(sel1=pop_select)\n",
    "    grp = pop_subset.groupby(\"population\")\n",
    "    grp_indices = grp.indices\n",
    "    ac_subpops = geno_p2_subset.count_alleles_subpops(grp_indices)\n",
    "    ac1 = ac_subpops[pop1]\n",
    "    ac2 = ac_subpops[pop2]\n",
    "    pos = calldata_phase2[chrom]['variants']['POS'][:]    \n",
    "    print (ac1.shape, ac2.shape, pos.shape, seq.shape)\n",
    "    dxy, windows, n_bases, counts = allel.windowed_divergence(pos, ac1, ac2, \n",
    "                                                                    size=window_size, \n",
    "                                                                    start=1, \n",
    "                                                                    stop=pos.max()\n",
    "                                                             )\n",
    "    x = np.mean(windows[n_bases >= min_n_bases], axis=1)\n",
    "    y = dxy[n_bases >= min_n_bases]\n",
    "    fig, ax = plt.subplots(figsize=(14, 4))\n",
    "    sns.despine(ax=ax, offset=10)\n",
    "    ax.plot(x, y, lw=.5)\n",
    "    ax.set_xlim(0, seq.size)\n",
    "    ax.set_title('%s vs %s (%s)' % (pop1, pop2, chrom), fontsize=14)\n",
    "    ax.set_ylabel('Dxy')\n",
    "    ax.set_xlabel('position')\n",
    "    ax.set_ylim(0,0.018)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "autosomes = '2R', '2L', '3R', '3L'\n",
    "chromosomes = autosomes + ('X',)\n",
    "\n",
    "\n",
    "class GenomeFigure(object):\n",
    "    \n",
    "    def __init__(self, genome, *args, **kwargs):\n",
    "        self.chromosomes = kwargs.pop('chromosomes', ['2R', '2L', '3R', '3L', 'X'])\n",
    "        maxchrsize = max(np.array(genome[chrom]).size for chrom in self.chromosomes)\n",
    "        fig = plt.figure(*args, **kwargs)\n",
    "        self.fig = fig\n",
    "        self.ax = dict()\n",
    "        for i, chrom in enumerate(self.chromosomes):\n",
    "            ax = fig.add_subplot(3, 2, i+1)\n",
    "            self.ax[chrom] = ax\n",
    "            S = np.array(genome[chrom])\n",
    "            if i % 2 == 1:\n",
    "                sns.despine(ax=ax, offset=10, top=True, left=True, right=False)\n",
    "                ax.set_xlim(0, maxchrsize)\n",
    "                ax.yaxis.tick_right()\n",
    "                ax.yaxis.set_label_position('right')\n",
    "            else:\n",
    "                ax.set_xlim((S.size)-(maxchrsize), S.size)\n",
    "                ax.yaxis.tick_left()\n",
    "                sns.despine(ax=ax, offset=10, top=True, left=False, right=True)\n",
    "            ax.set_xticks(range(0, S.size, int(5e6)))\n",
    "            ax.set_xticklabels(range(0, int(S.size/1e6), 5))\n",
    "            ax.set_title(chrom, fontweight='bold')\n",
    "            ax.xaxis.tick_bottom()\n",
    "        fig.tight_layout()\n",
    "        \n",
    "    def apply(self, f, **kwargs):\n",
    "        chromosomes = kwargs.pop('chromosomes', self.chromosomes)\n",
    "        for chrom in chromosomes:\n",
    "            ax = self.ax[chrom]\n",
    "            f(chrom, ax, **kwargs)\n",
    "        \n",
    "        \n",
    "def subplots(*args, **kwargs):\n",
    "    fig, ax = plt.subplots(*args, **kwargs)\n",
    "    sns.despine(ax=ax, offset=10)\n",
    "    return fig, ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_cache = dict()\n",
    "\n",
    "def load_ac(chrom, pop):\n",
    "    if (chrom, pop) in ac_cache:\n",
    "        return ac_cache[chrom, pop]\n",
    "    else:\n",
    "        if pop in out_species:\n",
    "            ac = calldata_out[chrom][pop]\n",
    "        else:\n",
    "            ac = calldata_biallel[chrom][pop]\n",
    "            ac_cache[chrom, pop] = ac\n",
    "        return ac\n",
    "    \n",
    "fourpop_cache = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def f4_analysis(chroms, A, B, C, D, regions=None, blen=100000, plot=False, ax=None):\n",
    "    \n",
    "    \n",
    "    \n",
    "    region_str = \",\".join([\"{0}_{1}\".format(r.start, r.stop) for r in regions])\n",
    "    \n",
    "    key = (\",\".join(chroms), region_str, A, B, C, D, blen)\n",
    "    if key in fourpop_cache:\n",
    "        # re-use from cache\n",
    "        d, d_se, d_z, d_vb, d_vj = fourpop_cache[key]\n",
    "        \n",
    "    else:\n",
    "        # load allele counts\n",
    "        aca = np.vstack([load_ac(chrom, A)[region] for chrom, region in zip(chroms, regions)])\n",
    "        acb = np.vstack([load_ac(chrom, B)[region] for chrom, region in zip(chroms, regions)])\n",
    "        acc = np.vstack([load_ac(chrom, C)[region] for chrom, region in zip(chroms, regions)])\n",
    "        acd = np.vstack([load_ac(chrom, D)[region] for chrom, region in zip(chroms, regions)])\n",
    "            \n",
    "        # run D test\n",
    "        d, d_se, d_z, d_vb, d_vj = allel.blockwise_patterson_d(aca, acb, acc, acd, \n",
    "                                                                     blen=blen)\n",
    "        \n",
    "        # cache for re-use\n",
    "        fourpop_cache[key] = d, d_se, d_z, d_vb, d_vj\n",
    "    \n",
    "    return d, d_se, d_z, d_vb, d_vj\n",
    "\n",
    "def f4_analys(As, Bs, Cs, Ds, chroms=chromosomes, regions=None, blen=100000):\n",
    "    \n",
    "    if regions is None:\n",
    "        regions = [slice(None)] * 4\n",
    "    \n",
    "    region_str = \",\".join([\"{0}_{1}\".format(r.start, r.stop) for r in regions])\n",
    "    # normalise inputs\n",
    "    if not isinstance(chroms, (list, tuple)):\n",
    "        chroms = [chroms]\n",
    "    if not isinstance(As, (list, tuple)):\n",
    "        As = [As]\n",
    "    if not isinstance(Bs, (list, tuple)):\n",
    "        Bs = [Bs]\n",
    "    if not isinstance(Cs, (list, tuple)):\n",
    "        Cs = [Cs]\n",
    "    if not isinstance(Ds, (list, tuple)):\n",
    "        Ds = [Ds]\n",
    "        \n",
    "    # setup output table\n",
    "    tbl = [['chromosome', 'test', 'D', 'SE', 'Z']]\n",
    "\n",
    "    for A in As:\n",
    "        for B in Bs:\n",
    "            if A != B:\n",
    "                for C in Cs:\n",
    "                    for D in Ds:\n",
    "                        if C != D:\n",
    "                            d, d_se, d_z, _, _ = f4_analysis(chroms, A, B, C, D, \n",
    "                                                             regions=regions, blen=blen)\n",
    "                            test = 'D(%s, %s; %s, %s)' % (A, B, C, D)\n",
    "                            row = [\",\".join(chroms), test, d, d_se, d_z]\n",
    "                            tbl += [row]\n",
    "    return tbl\n",
    "\n",
    "def tbl_display(tbl):\n",
    "        # display results\n",
    "    (etl\n",
    "     .wrap(tbl)\n",
    "     .interpolate('D', '%.3f')\n",
    "     .interpolate('SE', '%.4f')\n",
    "     .interpolate('Z', '%.1f')\n",
    "     .displayall(index_header=False, \n",
    "                 tr_style=lambda row: 'background-color: %s' % ('#afa' if float(row.Z) > 5 else '#aaf' if float(row.Z) < -5 else 'white')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_alleles_pos(chrom, start, stop):\n",
    "    \n",
    "    metadata = pd.read_csv(\"samples.meta.txt\", sep=\"\\t\")\n",
    "    pos_all = allel.SortedIndex(callset_biallel[chrom][\"variants/POS\"])\n",
    "    loc = pos_all.intersect_range(start, stop)\n",
    "    loc1, loc2 = pos_all.locate_intersection(loc)\n",
    "    geno = allel.GenotypeChunkedArray(callset_biallel[chrom][\"calldata/GT\"])\n",
    "    geno_subset = geno.subset(sel0=loc1)\n",
    "    ac = geno_subset.count_alleles()[:]\n",
    "    grp = metadata.groupby(\"population\")\n",
    "    grp_indices = grp.indices\n",
    "    ac_subpops = geno_subset.count_alleles_subpops(grp_indices)\n",
    "\n",
    "    return ac_subpops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tbl_d_plot(table, title, sort=None):\n",
    "    # plot\n",
    "    tbl = etl.wrap(table).sort(sort)\n",
    "    fig, (ax1, ax2) = plt.subplots(1,2,figsize=(20, tbl.nrows()*.2))\n",
    "    \n",
    "    ## First Plot - Bar Plot\n",
    "    x = tbl.values('D').array()\n",
    "    y = np.arange(tbl.nrows())\n",
    "    ax1.axvline(0, color='r', lw=1)\n",
    "    xerr = tbl.values('SE').array()\n",
    "    ax1.errorbar(x, y, xerr=xerr, fmt='o', lw=1, color='k', mew=1, mfc='k')\n",
    "    ax1.set_yticks(y)\n",
    "    ylbls = ['%s; Z = %.1f' % (v.test, v.Z) for v in tbl.records()]\n",
    "    ax1.set_yticklabels(ylbls)\n",
    "    ax1.yaxis.tick_right()\n",
    "    ax1.xaxis.tick_top()\n",
    "    ax1.xaxis.set_label_position('top')\n",
    "    ax1.set_xlim(-.2,.2)\n",
    "    ax1.set_ylim(-1, y.size)\n",
    "    ax1.grid(axis='both', color='#dddddd')\n",
    "    ax1.set_xlabel(\"Patterson's $D$ sorted by %s\" % (sort))\n",
    "    fig.suptitle('%s' % (title), y=1.02, fontsize=16)\n",
    "\n",
    "    ## Second plot - Heatmap\n",
    "    results = etl.todataframe(tbl)\n",
    "    df1 = results[['test', 'D','Z']].round(2)\n",
    "    heatmap1_data = pd.pivot_table(df1, values='Z', \n",
    "                     index=['test'], \n",
    "                     columns='D')\n",
    "    lm = sns.heatmap(heatmap1_data, cmap=\"RdBu\", vmin=15, vmax=-15, annot=True)\n",
    "    ax2 = lm.axes\n",
    "    ax2.set_xlim(-.7)\n",
    "    ax2.xaxis.set_label_position('top')\n",
    "    ax2.xaxis.tick_top()\n",
    "    plt.xticks(rotation=70)\n",
    "    ax2.set_xlabel(\"Patterson's $D$ sorted by population\")\n",
    "    plt.text(1.02, 0.5, 'Z-score', {'fontsize': 15},\n",
    "             horizontalalignment='left',\n",
    "             verticalalignment='center',\n",
    "             rotation=90,\n",
    "             clip_on=False,\n",
    "             transform=plt.gca().transAxes)\n",
    "\n",
    "    ax2.grid(False, 'major')\n",
    "    ax2.grid(True, 'minor')\n",
    "    ax2.set_yticks([t - 0.5 for t in ax2.get_yticks()], minor=True)\n",
    "    sns.despine()\n",
    "    fig.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------------------------------------------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:root] *",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
