{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Treemix Ag1000G phase2\n",
    "For build a dataset Treemix I need unlinked SNPs. So I have to prune my allele count datasets to obtain SNPs in high LD.\n",
    "For doing this I need:\n",
    "    - Phase2 Genotype callset\n",
    "    - Phase2 Allele count\n",
    "    - Outgroup Allele count\n",
    "\n",
    "In this notebook I edited an old Alistair's notebook of the Phase1 of Ag1000G (<b>20151001 treemix prep 4</b>). On my phase2 datasets I have already the biallelic allele counts so I skipped the searching and filtering for biallelic SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import my modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run imports.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import callsets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset_pass= callset_biallel\n",
    "allele_counts= zarr.open('data/phase2_biallel_allele_count.zarr/')\n",
    "outgroup_allele_counts= zarr.open('data/outgroup_alleles_phase2.zarr/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define functions to locate biallelic allele counts on a range for my outgroup and phase2 datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outgroup_ascertainment(chrom, start, stop, outgroups):\n",
    "    \n",
    "    # locate region\n",
    "    pos = allel.SortedIndex(callset_pass[chrom]['variants']['POS'][:])\n",
    "    locr = pos.locate_range(start, stop)\n",
    "    \n",
    "    # ascertain SNPs\n",
    "    loca = np.zeros(pos.shape, dtype='b1')\n",
    "    loca[locr] = True\n",
    "    log('outgroup ascertainment, initial', nnz(loca))\n",
    "    for s in outgroups:\n",
    "        ac = allel.AlleleCountsArray(outgroup_allele_counts[chrom][s][:])\n",
    "        # non-missing\n",
    "        locs = (ac.sum(axis=1) > 0)\n",
    "        loca &= locs\n",
    "        log(s, nnz(loca))\n",
    "        \n",
    "    return loca\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ingroup_ascertainment(chrom, start, stop, segpops):\n",
    "\n",
    "    # locate region\n",
    "    pos = allel.SortedIndex(callset_pass[chrom]['variants']['POS'][:])\n",
    "    locr = pos.locate_range(start, stop)\n",
    "\n",
    "    # ascertain SNPs\n",
    "    loca = np.zeros(pos.shape, dtype='b1')\n",
    "    loca[locr] = True\n",
    "    log('ingroup ascertainment, initial', nnz(loca))\n",
    "\n",
    "    \n",
    "    # require segregating\n",
    "    for pop in segpops:\n",
    "        ac = allel.AlleleCountsArray(allele_counts[chrom][pop][:])\n",
    "        loc_seg = ac.min(axis=1) > 0\n",
    "        loca &= loc_seg\n",
    "        log('after require segregating in', pop, nnz(loca))\n",
    "        \n",
    "    return loca"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define function for ld pruning. LD-pruning remove SNPs with an high correlation. Using windows this function compute pairwise LD between all SNPs within each window, then removing one SNP from each correlated pair."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define function for generating treemix file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_treemix(acs, fn):\n",
    "    pops = sorted(acs.keys())\n",
    "    n_variants = acs[pops[0]].shape[0]\n",
    "    n_alleles = acs[pops[0]].shape[1]\n",
    "    assert n_alleles == 2, 'only biallelic variants supported'\n",
    "    for pop in pops[1:]:\n",
    "        assert n_variants == acs[pop].shape[0], 'bad number of variants for pop %s' % pop\n",
    "        assert n_alleles == acs[pop].shape[1], 'bad number of alleles for pop %s' % pop\n",
    "        \n",
    "    with open(fn, 'wt', encoding='ascii') as f:\n",
    "        print(' '.join(pops), file=f)\n",
    "        for i in range(n_variants):\n",
    "            print(' '.join([','.join(map(str, acs[pop][i])) for pop in pops]), file=f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a new function that randomly downsample if I have a large dataset and applies ld-pruning on it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def downsample_and_prune(chrom, start, stop, loc_asc,\n",
    "                         n=100000, ldp_size=500, ldp_step=250, ldp_threshold=.1, ldp_n_iter=1):\n",
    "\n",
    "    # all variant positions\n",
    "    pos = allel.SortedIndex(callset_pass[chrom]['variants']['POS'][:])\n",
    "    posa = pos[loc_asc]\n",
    "\n",
    "    # randomly downsample\n",
    "    if n < posa.shape[0]:\n",
    "        posds = np.random.choice(posa, n, replace=False)\n",
    "        posds.sort()\n",
    "        posds = allel.SortedIndex(posds)\n",
    "    else:\n",
    "        # skip downsampling\n",
    "        posds = posa\n",
    "    locds = pos.locate_keys(posds)    \n",
    "\n",
    "    # load genotype data\n",
    "    genotype = allel.GenotypeChunkedArray(callset_pass[chrom]['calldata/GT'])\n",
    "    geno_subset = genotype.subset(sel0=loc_asc)\n",
    "    gn = geno_subset.to_n_alt()\n",
    "\n",
    "    \n",
    "    # prune    \n",
    "    for i in range(ldp_n_iter):\n",
    "        loc_unlinked = allel.locate_unlinked(gn, size=ldp_size, step=ldp_step, threshold=ldp_threshold)\n",
    "        n = np.count_nonzero(loc_unlinked)\n",
    "        n_remove = gn.shape[0] - n\n",
    "        log('iteration', i+1, 'retaining', n, 'removing', n_remove, 'variants')\n",
    "        gnu = gn.compress(loc_unlinked, axis=0)\n",
    "        posu = pos.compress(loc_unlinked)\n",
    "        locu = pos.locate_keys(posu)\n",
    "\n",
    "    return locu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define last function, the analysis function that includes all function below and applies these on my populations, outgroups, chromosomes of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_analysis(rname, chrom, start, stop, outgroups, segpops,\n",
    "                 n=100000, ldp_size=500, ldp_step=250, ldp_threshold=.1, ldp_n_iter=1):\n",
    "\n",
    "    # initial ascertainment\n",
    "    loc_og_asc = outgroup_ascertainment(chrom, start, stop, outgroups=outgroups)\n",
    "    loc_ig_asc = ingroup_ascertainment(chrom, start, stop, segpops=segpops)\n",
    "    loc_asc = loc_og_asc & loc_ig_asc\n",
    "    log('initial ascertainment', nnz(loc_asc))\n",
    "    \n",
    "    # downsample and prune\n",
    "    locu = downsample_and_prune(chrom, start, stop, loc_asc, \n",
    "                                n=n, ldp_size=ldp_size, ldp_step=ldp_step, \n",
    "                                ldp_threshold=ldp_threshold, ldp_n_iter=ldp_n_iter)\n",
    "    \n",
    "    # write allele counts\n",
    "    acsu = dict()\n",
    "    for pop in populations:\n",
    "        acsu[pop] = allele_counts[chrom][pop][:, :2][locu]\n",
    "    for pop in outgroups:\n",
    "        acsu[pop] = outgroup_allele_counts[chrom][pop][:, :2][locu]\n",
    "\n",
    "    outdir = 'd/data/treemix/seg_%s_og_%s_ldp_%s' % ('_'.join(segpops), '_'.join(outgroups), ldp_n_iter)\n",
    "    !mkdir -pv {outdir}\n",
    "    fn = os.path.join(outdir, '%s.allele_counts.txt' % rname)\n",
    "    to_treemix(acsu, fn)\n",
    "    !gzip -fv {fn}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Declaring values for generating my treemix file and ran on it for chromosome 3R, 3L, X, and the X region involved on speciation between <i>An.gambiae</i> and <i>An.coluzzii</i>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "outgroups = ['chri', 'arab', 'quad', 'meru', 'mela']\n",
    "segpops = ['AOcol', 'BFcol', 'CIcol', 'GHcol', 'GNcol','GHgam', 'CMgam', 'BFgam', 'GNgam', 'GQgam', 'UGgam', 'GAgam', 'FRgam','KE', 'GM', 'GW']\n",
    "n = 200000\n",
    "ldp_n_iter = 1\n",
    "region_3R_24mbp = '3R-24Mbp', '3R', 1, 24_000_000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----------------------------\n",
    "## Treemix on 24Mbp 3R-free"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3R-24Mbp 3R 1 24000000\n",
      "outgroup ascertainment, initial 5760020\n",
      "chri 4113161\n",
      "arab 4087191\n",
      "quad 4024544\n",
      "meru 3969408\n",
      "mela 3839974\n",
      "ingroup ascertainment, initial 5760020\n",
      "after require segregating in AOcol 627470\n",
      "after require segregating in BFcol 424553\n",
      "after require segregating in CIcol 373166\n",
      "after require segregating in GHcol 350110\n",
      "after require segregating in GNcol 181299\n",
      "after require segregating in GHgam 162341\n",
      "after require segregating in CMgam 162312\n",
      "after require segregating in BFgam 162127\n",
      "after require segregating in GNgam 161151\n",
      "after require segregating in GQgam 140491\n",
      "after require segregating in UGgam 140467\n",
      "after require segregating in GAgam 137080\n",
      "after require segregating in FRgam 91304\n",
      "after require segregating in KE 63790\n",
      "after require segregating in GM 63730\n",
      "after require segregating in GW 63725\n",
      "initial ascertainment 36930\n",
      "iteration 1 retaining 29203 removing 7727 variants\n",
      "d/data/treemix/seg_AOcol_BFcol_CIcol_GHcol_GNcol_GHgam_CMgam_BFgam_GNgam_GQgam_UGgam_GAgam_FRgam_KE_GM_GW_og_chri_arab_quad_meru_mela_ldp_1/3R-24Mbp.allele_counts.txt:\t 93.9% -- replaced with d/data/treemix/seg_AOcol_BFcol_CIcol_GHcol_GNcol_GHgam_CMgam_BFgam_GNgam_GQgam_UGgam_GAgam_FRgam_KE_GM_GW_og_chri_arab_quad_meru_mela_ldp_1/3R-24Mbp.allele_counts.txt.gz\n"
     ]
    }
   ],
   "source": [
    "rname, chrom, start, stop = region_3R_24mbp\n",
    "log(rname, chrom, start, stop)\n",
    "run_analysis(rname, chrom, start, stop, outgroups, segpops, n=n, ldp_n_iter=ldp_n_iter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AOcol</th>\n",
       "      <th>BFcol</th>\n",
       "      <th>BFgam</th>\n",
       "      <th>CIcol</th>\n",
       "      <th>CMgam</th>\n",
       "      <th>FRgam</th>\n",
       "      <th>GAgam</th>\n",
       "      <th>GHcol</th>\n",
       "      <th>GHgam</th>\n",
       "      <th>GM</th>\n",
       "      <th>...</th>\n",
       "      <th>GNgam</th>\n",
       "      <th>GQgam</th>\n",
       "      <th>GW</th>\n",
       "      <th>KE</th>\n",
       "      <th>UGgam</th>\n",
       "      <th>arab</th>\n",
       "      <th>chri</th>\n",
       "      <th>mela</th>\n",
       "      <th>meru</th>\n",
       "      <th>quad</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>181,1</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>178,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>223,1</td>\n",
       "      <td>24,0</td>\n",
       "      <td>0,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>155,1</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>109,1</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>137,1</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>156,0</td>\n",
       "      <td>138,12</td>\n",
       "      <td>10,174</td>\n",
       "      <td>135,7</td>\n",
       "      <td>31,563</td>\n",
       "      <td>0,46</td>\n",
       "      <td>20,118</td>\n",
       "      <td>105,5</td>\n",
       "      <td>1,23</td>\n",
       "      <td>33,97</td>\n",
       "      <td>...</td>\n",
       "      <td>3,77</td>\n",
       "      <td>1,17</td>\n",
       "      <td>60,122</td>\n",
       "      <td>89,7</td>\n",
       "      <td>7,217</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29198</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>593,1</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29199</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>223,1</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29200</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>183,1</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29201</th>\n",
       "      <td>139,17</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>141,1</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>1,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29202</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>183,1</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>...</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>0,1</td>\n",
       "      <td>8,0</td>\n",
       "      <td>20,0</td>\n",
       "      <td>20,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29203 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        AOcol   BFcol   BFgam  CIcol   CMgam FRgam   GAgam  GHcol GHgam  \\\n",
       "0       156,0   150,0   184,0  142,0   594,0  46,0   138,0  110,0  24,0   \n",
       "1       156,0   150,0   184,0  142,0   594,0  46,0   138,0  110,0  24,0   \n",
       "2       155,1   150,0   184,0  142,0   594,0  46,0   138,0  109,1  24,0   \n",
       "3       156,0   150,0   184,0  142,0   594,0  46,0   137,1  110,0  24,0   \n",
       "4       156,0  138,12  10,174  135,7  31,563  0,46  20,118  105,5  1,23   \n",
       "...       ...     ...     ...    ...     ...   ...     ...    ...   ...   \n",
       "29198   156,0   150,0   184,0  142,0   593,1  48,0   138,0  110,0  24,0   \n",
       "29199   156,0   150,0   184,0  142,0   594,0  48,0   138,0  110,0  24,0   \n",
       "29200   156,0   150,0   183,1  142,0   594,0  48,0   138,0  110,0  24,0   \n",
       "29201  139,17   150,0   184,0  141,1   594,0  48,0   138,0  110,0  24,0   \n",
       "29202   156,0   150,0   183,1  142,0   594,0  48,0   138,0  110,0  24,0   \n",
       "\n",
       "          GM  ... GNgam GQgam      GW    KE  UGgam  arab chri mela  meru  quad  \n",
       "0      130,0  ...  80,0  18,0   181,1  96,0  224,0  24,0  1,0  8,0  20,0  20,0  \n",
       "1      130,0  ...  80,0  18,0   178,0  96,0  223,1  24,0  0,0  8,0  20,0  20,0  \n",
       "2      130,0  ...  80,0  18,0   182,0  96,0  224,0  24,0  1,0  8,0  20,0  20,0  \n",
       "3      130,0  ...  80,0  18,0   182,0  96,0  224,0  24,0  1,0  8,0  20,0  20,0  \n",
       "4      33,97  ...  3,77  1,17  60,122  89,7  7,217  24,0  1,0  8,0  20,0  20,0  \n",
       "...      ...  ...   ...   ...     ...   ...    ...   ...  ...  ...   ...   ...  \n",
       "29198  130,0  ...  80,0  18,0   182,0  96,0  224,0  24,0  1,0  8,0  20,0  20,0  \n",
       "29199  130,0  ...  80,0  18,0   182,0  96,0  223,1  24,0  1,0  8,0  20,0  20,0  \n",
       "29200  130,0  ...  80,0  18,0   182,0  96,0  224,0  24,0  1,0  8,0  20,0  20,0  \n",
       "29201  130,0  ...  80,0  18,0   182,0  96,0  224,0  24,0  1,0  8,0  20,0  20,0  \n",
       "29202  130,0  ...  80,0  18,0   182,0  96,0  224,0  24,0  0,1  8,0  20,0  20,0  \n",
       "\n",
       "[29203 rows x 21 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('d/data/treemix/seg_AOcol_BFcol_CIcol_GHcol_GNcol_GHgam_CMgam_BFgam_GNgam_GQgam_UGgam_GAgam_FRgam_KE_GM_GW_og_chri_arab_quad_meru_mela_ldp_1/3R-24Mbp.allele_counts.txt.gz', sep = ' ')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AOcol</th>\n",
       "      <th>BFcol</th>\n",
       "      <th>BFgam</th>\n",
       "      <th>CIcol</th>\n",
       "      <th>CMgam</th>\n",
       "      <th>FRgam</th>\n",
       "      <th>GAgam</th>\n",
       "      <th>GHcol</th>\n",
       "      <th>GHgam</th>\n",
       "      <th>GM</th>\n",
       "      <th>GNcol</th>\n",
       "      <th>GNgam</th>\n",
       "      <th>GQgam</th>\n",
       "      <th>GW</th>\n",
       "      <th>KE</th>\n",
       "      <th>UGgam</th>\n",
       "      <th>chri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>181,1</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>155,1</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>109,1</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>46,0</td>\n",
       "      <td>137,1</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>156,0</td>\n",
       "      <td>138,12</td>\n",
       "      <td>10,174</td>\n",
       "      <td>135,7</td>\n",
       "      <td>31,563</td>\n",
       "      <td>0,46</td>\n",
       "      <td>20,118</td>\n",
       "      <td>105,5</td>\n",
       "      <td>1,23</td>\n",
       "      <td>33,97</td>\n",
       "      <td>8,0</td>\n",
       "      <td>3,77</td>\n",
       "      <td>1,17</td>\n",
       "      <td>60,122</td>\n",
       "      <td>89,7</td>\n",
       "      <td>7,217</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>593,1</td>\n",
       "      <td>46,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29198</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>593,1</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29199</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>223,1</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29200</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>183,1</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29201</th>\n",
       "      <td>139,17</td>\n",
       "      <td>150,0</td>\n",
       "      <td>184,0</td>\n",
       "      <td>141,1</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>1,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29202</th>\n",
       "      <td>156,0</td>\n",
       "      <td>150,0</td>\n",
       "      <td>183,1</td>\n",
       "      <td>142,0</td>\n",
       "      <td>594,0</td>\n",
       "      <td>48,0</td>\n",
       "      <td>138,0</td>\n",
       "      <td>110,0</td>\n",
       "      <td>24,0</td>\n",
       "      <td>130,0</td>\n",
       "      <td>8,0</td>\n",
       "      <td>80,0</td>\n",
       "      <td>18,0</td>\n",
       "      <td>182,0</td>\n",
       "      <td>96,0</td>\n",
       "      <td>224,0</td>\n",
       "      <td>0,1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19211 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        AOcol   BFcol   BFgam  CIcol   CMgam FRgam   GAgam  GHcol GHgam  \\\n",
       "0       156,0   150,0   184,0  142,0   594,0  46,0   138,0  110,0  24,0   \n",
       "2       155,1   150,0   184,0  142,0   594,0  46,0   138,0  109,1  24,0   \n",
       "3       156,0   150,0   184,0  142,0   594,0  46,0   137,1  110,0  24,0   \n",
       "4       156,0  138,12  10,174  135,7  31,563  0,46  20,118  105,5  1,23   \n",
       "5       156,0   150,0   184,0  142,0   593,1  46,0   138,0  110,0  24,0   \n",
       "...       ...     ...     ...    ...     ...   ...     ...    ...   ...   \n",
       "29198   156,0   150,0   184,0  142,0   593,1  48,0   138,0  110,0  24,0   \n",
       "29199   156,0   150,0   184,0  142,0   594,0  48,0   138,0  110,0  24,0   \n",
       "29200   156,0   150,0   183,1  142,0   594,0  48,0   138,0  110,0  24,0   \n",
       "29201  139,17   150,0   184,0  141,1   594,0  48,0   138,0  110,0  24,0   \n",
       "29202   156,0   150,0   183,1  142,0   594,0  48,0   138,0  110,0  24,0   \n",
       "\n",
       "          GM GNcol GNgam GQgam      GW    KE  UGgam chri  \n",
       "0      130,0   8,0  80,0  18,0   181,1  96,0  224,0  1,0  \n",
       "2      130,0   8,0  80,0  18,0   182,0  96,0  224,0  1,0  \n",
       "3      130,0   8,0  80,0  18,0   182,0  96,0  224,0  1,0  \n",
       "4      33,97   8,0  3,77  1,17  60,122  89,7  7,217  1,0  \n",
       "5      130,0   8,0  80,0  18,0   182,0  96,0  224,0  1,0  \n",
       "...      ...   ...   ...   ...     ...   ...    ...  ...  \n",
       "29198  130,0   8,0  80,0  18,0   182,0  96,0  224,0  1,0  \n",
       "29199  130,0   8,0  80,0  18,0   182,0  96,0  223,1  1,0  \n",
       "29200  130,0   8,0  80,0  18,0   182,0  96,0  224,0  1,0  \n",
       "29201  130,0   8,0  80,0  18,0   182,0  96,0  224,0  1,0  \n",
       "29202  130,0   8,0  80,0  18,0   182,0  96,0  224,0  0,1  \n",
       "\n",
       "[19211 rows x 17 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_3_chri = df.drop(df.columns[[16,18,19,20]], axis=1)\n",
    "chrom_3_chri = chrom_3_chri[chrom_3_chri.chri != '0,0']\n",
    "chrom_3_chri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_3_chri.to_csv('chrom_3L_chri.txt.gz',index=False, sep=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
