{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ultil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "General purpose utility functions.\n",
    "\n",
    "\"\"\"\n",
    "from __future__ import absolute_import, print_function, division\n",
    "\n",
    "\n",
    "# standard library imports\n",
    "import contextlib\n",
    "import sys\n",
    "import datetime\n",
    "\n",
    "\n",
    "# third party library imports\n",
    "import humanize\n",
    "import pandas\n",
    "import petl as etl\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "_slog_indent = -2\n",
    "\n",
    "\n",
    "def log(*msg):\n",
    "    \"\"\"Simple logging function that flushes immediately to stdout.\"\"\"\n",
    "    s = ' '.join(map(str, msg))\n",
    "    print(s, file=sys.stdout)\n",
    "    sys.stdout.flush()\n",
    "\n",
    "\n",
    "@contextlib.contextmanager\n",
    "def timer(*msg):\n",
    "    before = datetime.datetime.now()\n",
    "    try:\n",
    "        yield\n",
    "    except:\n",
    "        after = datetime.datetime.now()\n",
    "        elapsed = (after - before).total_seconds()\n",
    "        done = 'errored after %s' % humanize.naturaldelta(elapsed)\n",
    "        if not msg:\n",
    "            msg = done\n",
    "        else:\n",
    "            msg = ', '.join(map(str, msg)) + ', ' + done\n",
    "        print(msg, file=sys.stderr)\n",
    "        sys.stderr.flush()\n",
    "        raise\n",
    "    else:\n",
    "        after = datetime.datetime.now()\n",
    "        elapsed = (after - before).total_seconds()\n",
    "        done = 'done in %s' % humanize.naturaldelta(elapsed)\n",
    "        if not msg:\n",
    "            msg = done\n",
    "        else:\n",
    "            msg = ', '.join(map(str, msg)) + ', ' + done\n",
    "        print(msg, file=sys.stdout)\n",
    "        sys.stdout.flush()\n",
    "\n",
    "\n",
    "@contextlib.contextmanager\n",
    "def section(*title):\n",
    "    global _slog_indent\n",
    "    before = datetime.datetime.now()\n",
    "    _slog_indent += 2\n",
    "    prefix = (' ' * _slog_indent) + '[' + ', '.join(map(str, title)) + '] '\n",
    "\n",
    "    def slog(*msg, file=sys.stdout):\n",
    "        print(prefix + ' '.join(map(str, msg)), file=file)\n",
    "        file.flush()\n",
    "\n",
    "    slog('begin')\n",
    "\n",
    "    try:\n",
    "        yield slog\n",
    "\n",
    "    except:\n",
    "        after = datetime.datetime.now()\n",
    "        elapsed = (after - before).total_seconds()\n",
    "        msg = 'errored after %s' % humanize.naturaldelta(elapsed)\n",
    "        slog(msg, file=sys.stderr)\n",
    "        _slog_indent -= 2\n",
    "        raise\n",
    "\n",
    "    else:\n",
    "        after = datetime.datetime.now()\n",
    "        elapsed = (after - before).total_seconds()\n",
    "        msg = 'done in %s' % humanize.naturaldelta(elapsed)\n",
    "        slog(msg, file=sys.stdout)\n",
    "        _slog_indent -= 2\n",
    "\n",
    "\n",
    "def _h5ls(h5o, currentdepth, maxdepth, maxitems, prefix):\n",
    "    if maxdepth is not None and currentdepth == maxdepth:\n",
    "        return\n",
    "    for i, k in enumerate(h5o.keys()):\n",
    "        path = prefix + '/' + k\n",
    "        if maxitems is not None and i == maxitems:\n",
    "            print(prefix + '/...')\n",
    "            break\n",
    "        v = h5o[k]\n",
    "        print(path + ' : ' + repr(v))\n",
    "        if hasattr(v, 'keys'):\n",
    "            _h5ls(v, currentdepth+1, maxdepth=maxdepth, maxitems=maxitems, prefix=path)\n",
    "\n",
    "\n",
    "def h5ls(h5o, maxdepth=None, maxitems=None):\n",
    "    \"\"\"Obtain a recursive listing of the contents of an HDF5 file or group.\"\"\"\n",
    "    _h5ls(h5o, 0, maxdepth=maxdepth, maxitems=maxitems, prefix='')\n",
    "\n",
    "\n",
    "def fig_linear_genome(plotf, genome, chromosomes=('2R', '2L', '3R', '3L', 'X'),\n",
    "                      fig=None, bottom=0, height=1, width_factor=1.08, chrom_pad=0.035,\n",
    "                      clip_patch_kwargs=None, **kwargs):\n",
    "    \"\"\"Utility function to make a linear genome figure.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    plotf : function\n",
    "        Function to plot a single chromosome. Must accept 'chrom' and 'ax' arguments and also\n",
    "        flexible **kwargs.\n",
    "    genome : pyfasta.Fasta\n",
    "        Reference sequence. Used to compute genome and chromosome sizes.\n",
    "    chromosomes : tuple of strings, optional\n",
    "        Chromosomes to plot.\n",
    "    fig : figure, optional\n",
    "        Figure to plot on. If not provided, a new figure will be created.\n",
    "    bottom : float, optional\n",
    "        Figure coordinate to position bottom of axes.\n",
    "    height : float, optional\n",
    "        Figure height to use for axes.\n",
    "    width_factor : float, optional\n",
    "        Used to scale width of each chromosome subplot.\n",
    "    chrom_pad : float, optional\n",
    "        Used to set padding between chromosomes.\n",
    "    clip_patch_kwargs : dict-like, optional\n",
    "        Arguments for the clip path if used.\n",
    "    **kwargs\n",
    "        Passed through to `plotf`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    axs : dict\n",
    "        A dictionary mapping chromosome names onto axes objects.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    from matplotlib.path import Path\n",
    "\n",
    "    # compute assembled genome size\n",
    "    genome_size = sum(len(genome[chrom]) for chrom in chromosomes)\n",
    "\n",
    "    # setup figure\n",
    "    if fig is None:\n",
    "        fig = plt.figure(figsize=(8, 1))\n",
    "\n",
    "    # setup clip patch\n",
    "    if clip_patch_kwargs is None:\n",
    "        clip_patch_kwargs = dict()\n",
    "    clip_patch_kwargs.setdefault('edgecolor', 'k')\n",
    "    clip_patch_kwargs.setdefault('facecolor', 'none')\n",
    "    clip_patch_kwargs.setdefault('lw', 1)\n",
    "\n",
    "    # setup axes\n",
    "    left = 0\n",
    "    axs = dict()\n",
    "\n",
    "    for chrom in chromosomes:\n",
    "\n",
    "        # calculate width needed for this chrom\n",
    "        width = len(genome[chrom]) / (genome_size * width_factor)\n",
    "\n",
    "        # create axes\n",
    "        ax = fig.add_axes([left, bottom, width, height])\n",
    "        ax.set_facecolor((1, 1, 1, 0))\n",
    "        axs[chrom] = ax\n",
    "\n",
    "        # construct clip path\n",
    "        if chrom in {'2R', '3R'}:\n",
    "            verts = [(0.01, 0.02), (0.9, 0.02), (1.01, 0.3), (1.01, 0.7), (0.9, .98), (0.01, .98), (0.01, 0.02)]\n",
    "            codes = [Path.MOVETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY]\n",
    "        elif chrom == \"X\":\n",
    "            verts = [(0.01, 0.02), (0.9, 0.02), (0.99, 0.3), (0.99, 0.7), (0.9, .98), (0.01, .98), (0.01, 0.02)]\n",
    "            codes = [Path.MOVETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY]\n",
    "        else:\n",
    "            verts = [(0.1, 0.02), (.99, 0.02), (.99, .98), (.1, .98), (-0.01, .7), (-0.01, .3), (0.1, 0.02)]\n",
    "            codes = [Path.MOVETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.LINETO, Path.CLOSEPOLY]\n",
    "        path = Path(verts, codes)\n",
    "        clip_patch = mpl.patches.PathPatch(path, transform=ax.transAxes, **clip_patch_kwargs)\n",
    "\n",
    "        # do the plotting\n",
    "        plotf(chrom=chrom, ax=ax, clip_patch=clip_patch, **kwargs)\n",
    "\n",
    "        # increment left coordinate\n",
    "        left += len(genome[chrom]) / (genome_size * width_factor)\n",
    "        if chrom in {'2L', '3L'}:\n",
    "            left += chrom_pad\n",
    "\n",
    "    return axs\n",
    "\n",
    "\n",
    "def geneset_to_pandas(geneset):\n",
    "    \"\"\"Life is a bit easier when a geneset is a pandas DataFrame.\"\"\"\n",
    "    items = []\n",
    "    for n in geneset.dtype.names:\n",
    "        v = geneset[n]\n",
    "        # convert bytes columns to unicode (which pandas then converts to object)\n",
    "        if v.dtype.kind == 'S':\n",
    "            v = v.astype('U')\n",
    "        items.append((n, v))\n",
    "    return pandas.DataFrame.from_items(items)\n",
    "\n",
    "\n",
    "class SeqFeature(object):\n",
    "    \"\"\"Genomic sequence feature, with utilities for mapping between coordinate systems.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    seqid : string\n",
    "        Chromosome or contig.\n",
    "    start : int\n",
    "        Start coordinate, 1-based.\n",
    "    end : int\n",
    "        End coordinate, 1-based, end-inclusive.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, seqid, start, end, strand=None, genome=None, label=None):\n",
    "        self.seqid = seqid\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "        self.genome = genome\n",
    "        self.label = label\n",
    "\n",
    "    @property\n",
    "    def loc0(self):\n",
    "        \"\"\"A zero-based stop-exclusive slice.\"\"\"\n",
    "        return slice(self.start - 1, self.end)\n",
    "\n",
    "    @property\n",
    "    def query_str(self):\n",
    "        \"\"\"A pandas-style query string.\"\"\"\n",
    "        return \"(seqid == %r) & (start >= %s) & (end <= %s)\" % (self.seqid, self.start, self.end)\n",
    "\n",
    "    @property\n",
    "    def region_str(self):\n",
    "        \"\"\"A samtools-style region string.\"\"\"\n",
    "        return \"%s:%s-%s\" % (self.seqid, self.start, self.end)\n",
    "\n",
    "    @property\n",
    "    def seq(self):\n",
    "        \"\"\"The reference sequence.\"\"\"\n",
    "        return self.genome[self.seqid][self.loc0]\n",
    "\n",
    "    def __len__(self):\n",
    "        # include start and end positions in length\n",
    "        return self.end - self.start + 1\n",
    "\n",
    "    def __iter__(self):\n",
    "        yield self.seqid\n",
    "        yield self.start\n",
    "        yield self.end\n",
    "\n",
    "    def __repr__(self):\n",
    "        r = '<%s' % type(self).__name__\n",
    "        if self.label:\n",
    "            r += ' %r' % self.label\n",
    "        r += ' ' + self.region_str\n",
    "        if self.strand:\n",
    "            r += ' (%s)' % self.strand\n",
    "        r += '>'\n",
    "        return r\n",
    "\n",
    "\n",
    "chromosomes = '2R', '2L', '3R', '3L', 'X', 'Y_unplaced', 'UNKN'\n",
    "autosomes = chromosomes[:4]\n",
    "\n",
    "\n",
    "gene_labels = {\n",
    "    'AGAP009195': 'Gste1',\n",
    "    'AGAP009194': 'Gste2',\n",
    "    'AGAP009197': 'Gste3',\n",
    "    'AGAP009193': 'Gste4',\n",
    "    'AGAP009192': 'Gste5',\n",
    "    'AGAP009191': 'Gste6',\n",
    "    'AGAP009196': 'Gste7',\n",
    "    'AGAP009190': 'Gste8',\n",
    "    'AGAP004707': 'Vgsc',\n",
    "    'AGAP002862': 'Cyp6aa1',\n",
    "    'AGAP013128': 'Cyp6aa2',\n",
    "    'AGAP002863': 'Coeae6o',\n",
    "    'AGAP002865': 'Cyp6p3',\n",
    "    'AGAP002866': 'Cyp6p5',\n",
    "    'AGAP002867': 'Cyp6p4',\n",
    "    'AGAP002868': 'Cyp6p1',\n",
    "    'AGAP002869': 'Cyp6p2',\n",
    "    'AGAP002870': 'Cyp6ad1',\n",
    "    'AGAP002915': 'Pcsk4/furin',\n",
    "    'AGAP002825': 'Pp01',\n",
    "    'AGAP002824': 'Gprtak1',\n",
    "    'AGAP006028': 'Gaba',\n",
    "    'AGAP010815': 'Tep1'\n",
    "}\n",
    "\n",
    "\n",
    "def get_geneset_features(geneset_fn, chrom, start=None, stop=None):\n",
    "    \"\"\"Function to load geneset features for a specific genome region via petl.\"\"\"\n",
    "    if start and stop:\n",
    "        region = '%s:%s-%s' % (chrom, start, stop)\n",
    "    else:\n",
    "        region = chrom\n",
    "    return etl.fromgff3(geneset_fn, region=region)\n",
    "\n",
    "\n",
    "def plot_genes(genome, geneset_fn, chrom, start=1, stop=None, ax=None, height=.3, label=False,\n",
    "               labels=None, label_unnamed=True, label_rotation=45, barh_kwargs=None):\n",
    "\n",
    "    if stop is None:\n",
    "        stop = len(genome[chrom])\n",
    "\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(7, 1))\n",
    "        sns.despine(ax=ax, offset=5)\n",
    "\n",
    "    genes = get_geneset_features(geneset_fn, chrom, start, stop).eq('type', 'gene').records()\n",
    "\n",
    "    fwd_ranges = [(g.start, (g.end - g.start)) for g in genes if g.strand == '+']\n",
    "    rev_ranges = [(g.start, (g.end - g.start)) for g in genes if g.strand == '-']\n",
    "    if barh_kwargs is None:\n",
    "        barh_kwargs = dict()\n",
    "    barh_kwargs.setdefault('color', 'k')\n",
    "    ax.broken_barh(fwd_ranges, (.5, height), **barh_kwargs)\n",
    "    ax.broken_barh(rev_ranges, (.5-height, height), **barh_kwargs)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.axhline(.5, color='k', linestyle='-')\n",
    "    ax.set_xlim(start, stop)\n",
    "    ax.set_yticks([.5-(height/2), .5+(height/2)])\n",
    "    ax.set_yticklabels(['-', '+'])\n",
    "    ax.set_ylabel('genes', rotation=0, ha='right', va='center')\n",
    "\n",
    "    if label:\n",
    "        for gene in genes:\n",
    "            gid = gene.attributes['ID']\n",
    "            if labels and gid not in labels and not label_unnamed:\n",
    "                continue\n",
    "            if labels and gid in labels:\n",
    "                label = labels[gid]\n",
    "            else:\n",
    "                label = gid\n",
    "            x = gene.start\n",
    "            if x < start:\n",
    "                x = start\n",
    "            if x > stop:\n",
    "                x = stop\n",
    "            if gene.strand == '+':\n",
    "                rotation = label_rotation\n",
    "                y = .5 + height\n",
    "                ax.text(x, y, label, rotation=rotation, fontsize=10, ha='left', va='bottom')\n",
    "            else:\n",
    "                rotation = -label_rotation\n",
    "                y = .5 - height\n",
    "                ax.text(x, y, label, rotation=rotation, fontsize=10, ha='left', va='top')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
